vm.h

```diff
@@ -31,16 +32,52 @@ typedef struct {
     vm_handler handler;
 } vm_opcode_impl;
 
+typedef struct {
+    enum { SYMBOLIC, NUMERIC } type;
+    char *label;
+    int next_pc;
+} vm_label;
+
 #define VM_T(_op) _op->type
 #define VM_INT(_op) _op->value.vint
 
-typedef struct __vm_env vm_env;
+/* Constant pool max size */
+#define CPOOL_MAX_SIZE 100
+
+/* Instruction max size */
+#define INSTS_MAX_SIZE 200
+
+/* Temporary storage max size */
+#define TEMPS_MAX_SIZE 150
+
+/* OPCODE impl max size */
+#define OPCODE_IMPL_MAX_SIZE 256
+
+/* Labels max size */
+#define LABELS_MAX_SIZE 256
+
+typedef struct __vm_env {
+    vm_inst insts[INSTS_MAX_SIZE];                /* Program instructions */
+    vm_value cpool[CPOOL_MAX_SIZE];               /* Constant pool */
+    vm_value temps[TEMPS_MAX_SIZE];               /* Temporary storage */
+    vm_opcode_impl impl[OPCODE_IMPL_MAX_SIZE];    /* OPCODE impl */
+    vm_label labels[LABELS_MAX_SIZE];             /* Labels */
+    vm_operand *label_reference[LABELS_MAX_SIZE]; /* Operand's label reference */
+    int insts_count;
+    int cpool_count;
+    int temps_count;
+    int labels_count;
+    int label_reference_count;
+} vm_env;
 
 vm_env *vm_new();
 void vm_free(vm_env *);
 
 size_t vm_add_const(vm_env *, int, void *);
 size_t vm_add_inst(vm_env *, vm_inst);
+size_t vm_add_label(vm_env *, const int, char *);
+int vm_find_label(vm_env *, const char *);
+void vm_hook_label(vm_env *);
```

Comment:
`vm_hook_label` is not intuitive. Using `register` or `insert` might be a better alternative.
Terms: http://arantxa.ii.uam.es/~modonnel/Compilers/04_SymbolTablesI.pdf